arch/arm/core/cortex_m/mpu/arm_mpu.c

```diff
@@ -72,6 +72,9 @@ void arm_core_mpu_enable(void)
  */
 void arm_core_mpu_disable(void)
 {
+	/* Force any outstanding transfers to complete before disabling MPU */
+	__DMB();
```

Comment:
In principle this is OK, but may introduce performance issues. Is it needed for proper cache operation, or do you think that this, in general, is always needed?

Reply:
Also: would we need an instruction barrier, too?

Reply:
I have found that adding the memory barrier helps to pass a few tests from the mem_protect testsuite. This code is common to both ARMv7-M and ARMv8-M, and the ARMv8-M clearly says this **must** be done before disabling the MPU. And this is in line with what is done for enabling the MPU. Not using a memory barrier means some load/store might be done with the MPU disabled, so with a different memory permissions and possibly caching modes.

I am not sure about the instruction barrier here. It has much more performances impact, than a memory barrier and it seems to be one of your concern. Adding one does not change anything to the tests result and it is not mentioned in the ARM documentation. However CMSIS 5 uses `__DSB()` followed by `__ISB()` before disabling the MPU for both ARMv7 and ARMv8.

Reply:
> I am not sure about the instruction barrier here. It has much more performances impact, than a memory barrier and it seems to be one of your concern. Adding one does not change anything to the tests result and it is not mentioned in the ARM documentation. However CMSIS 5 uses __DSB() followed by __ISB() before disabling the MPU for both ARMv7 and ARMv8.

Exactly. I guess we can think of adding the ISB if we see the need to do so.